extensions/src/main/java/org/pentaho/platform/plugin/services/metadata/SessionCachingMetadataDomainRepository.java

```diff
@@ -368,27 +370,63 @@ private void removeDomainFromIDCache( String domainId ) {
     }
   }
 
+  /**
+   * Remove a single Multiple ID from all session domain ID caches
+   *
+   * @param domains
+   */
+  private void removeDomainsFromIDCache( List<String> domains ) {
+    try {
+      Set<?> cachedObjects = cacheManager.getAllKeysFromRegionCache( CACHE_REGION );
+      if ( cachedObjects != null ) {
+        for ( Object k : cachedObjects ) {
+          if ( k instanceof String ) {
+            String key = (String) k;
+            if ( key != null && key.startsWith( DOMAIN_CACHE_KEY_PREDICATE ) ) {
+
+              for ( String domainId : domains ) {
+                Set<String> domainIds = (Set<String>) cacheManager.getFromRegionCache( CACHE_REGION, key );
+                synchronized ( domainIds ) {
+                  domainIds.remove( domainId );
+                  cacheManager.putInRegionCache( CACHE_REGION, key, domainIds );
+                }
+              }
+            }
+          }
+        }
+      }
+    } catch ( Throwable e ) {
+      // due to a known issue in hibernate cache
+      // the getAll* methods of ICacheManager throw a NullPointerException if
+      // cache values are null (this can happen due to cache object timeouts)
+      // please see: http://opensource.atlassian.com/projects/hibernate/browse/HHH-3248
+      if ( logger.isDebugEnabled() ) {
+        logger.debug( "", e ); //$NON-NLS-1$
+      }
+    }
+  }
+
   protected String generateDomainIdCacheKeyForSession( IPentahoSession session ) {
     return DOMAIN_CACHE_KEY_PREDICATE + session.getId();
   }
 
   @Override
-  public void removeDomain( final String domainId ) {
+  public synchronized void removeDomain( final String domainId ) {
```

Comment:
By definition, this class is session based ( SessionCachingMetadataDomainRepository ). However, this change is synchronizing on the entire MetadataRepository instance. So removing a domain for one user in their own session ends up blocking another user in another session from storing their domain or removing their model.  I don't understand why this should be done at this level. If there really is synchronization required somewhere, I don't suspect it's here. What specifically happens in the multi-threaded testing without these three methods being synchronized that this solves ?